pkg/linters/components/compimpl/linter.go

```diff
@@ -0,0 +1,143 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package compimpl is a linter that checks if the component implementation follows the expected naming convention.
+package compimpl
+
+import (
+	"fmt"
+	"go/ast"
+	"os"
+	"regexp"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+func init() {
+	RegisterLinter("compimpl", "Check the component implementation follows the naming convention", run)
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	r := regexp.MustCompile(".*comp/.*/.*/.*impl/.*")
+	for _, f := range pass.Files {
+		implFolder, err := doesFileImplementComp(pass, f, r)
+		if !implFolder {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		structTypeNames := getExportedStructTypeNames(f)
+		if !structTypeNames["Provides"] {
+			pass.Report(analysis.Diagnostic{
+				Pos:     f.Pos(),
+				Message: missingProvides(),
+			})
+		}
+
+		if !structTypeNames["Requires"] {
+			pass.Report(analysis.Diagnostic{
+				Pos:     f.Pos(),
+				Message: missingRequires(),
+			})
+		}
+		if !hasComponentConstructor(f) {
+			pass.Report(analysis.Diagnostic{
+				Pos:     f.Pos(),
+				Message: missingConstructor(),
+			})
+		}
+	}
+
+	return nil, nil
+}
+
+func doesFileImplementComp(pass *analysis.Pass, file *ast.File, r *regexp.Regexp) (bool, error) {
+	path := pass.Fset.Position(file.Pos()).Filename
+	if r.MatchString(path) {
+		content, err := os.ReadFile(path)
+		if err != nil {
+			return false, err
+		}
+		return strings.Contains(string(content), "func Module() fxutil.Module {"), nil
```

Comment:
It seems odd to me that a Go linter reads files and checks that they contains strings, instead of going down the AST

Reply:
I have changed the way the linter works as there is no `Module` function outside `fx` folder.